File: src/main/java/schemabuilder/processor/pipelines/parsing/scalars/GraphQLScalarParser.java
package schemabuilder.processor.pipelines.parsing.scalars;

import graphql.schema.GraphQLScalarType;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import schemabuilder.annotations.graphql.GraphQLScalar;
import schemabuilder.annotations.graphql.GraphQLSchemaConfiguration;
import schemabuilder.processor.pipelines.parsing.GraphQLClassParserStrategy;
import schemabuilder.processor.wiring.InstanceFetcher;

public class GraphQLScalarParser implements GraphQLClassParserStrategy {

    private GraphQLScalarBank scalarBank = GraphQLScalarBank.getInstance();

    @Override
    public void parse(Class<?> clazz, InstanceFetcher fetcher) {
        if (!clazz.isAnnotationPresent(GraphQLSchemaConfiguration.class)) {
            return;
        }

        Object instance = fetcher.getInstance(clazz);

        for (Method method : clazz.getDeclaredMethods()) {
            GraphQLScalar annotation = method.getAnnotation(GraphQLScalar.class);
            if(annotation == null || !method.getReturnType().equals(graphql.schema.GraphQLScalarType.class)) {
                continue;
            }

            method.setAccessible(true);

            try {
                scalarBank.addScalar(new schemabuilder.processor.pipelines.parsing.scalars.GraphQLScalarType(annotation.value(), (GraphQLScalarType) method.invoke(instance)));
            } catch (InvocationTargetException | IllegalAccessException e) {
                e.printStackTrace();
            }
        }
    }
}

File: src/main/java/schemabuilder/processor/pipelines/parsing/scalars/GraphQLScalarParser.kt
package schemabuilder.processor.pipelines.parsing.scalars

import schemabuilder.annotations.graphql.GraphQLScalar
import schemabuilder.annotations.graphql.GraphQLSchemaConfiguration
import schemabuilder.processor.pipelines.parsing.GraphQLClassParserStrategy
import schemabuilder.processor.pipelines.parsing.ParsedResults
import schemabuilder.processor.wiring.InstanceFetcher
import kotlin.reflect.KClass
import kotlin.reflect.full.declaredFunctions
import kotlin.reflect.full.findAnnotation
import kotlin.reflect.full.functions
import kotlin.reflect.full.hasAnnotation
import kotlin.reflect.jvm.isAccessible

class GraphQLScalarParser : GraphQLClassParserStrategy {

    override fun parse(clazz: KClass<*>, fetcher: InstanceFetcher) {
        for (method in clazz.declaredFunctions) {
            val annotation = method.findAnnotation<GraphQLScalar>()

            try {
                method.isAccessible = true
                ParsedResults.scalars.add(GraphQLScalarType(annotation?.value!!, method.call() as graphql.schema.GraphQLScalarType))

            } catch (e: Exception) {

            }
        }
    }

    @ExperimentalStdlibApi
    override fun shouldParse(clazz: KClass<*>) = clazz.hasAnnotation<GraphQLSchemaConfiguration>()
}